### DISCUSSION TITLE
Separation between user and "admin" Quota, Settings, etc.

### DISCUSSION THREAD

**Discussion by LumiWasTaken** [Kilvoctu/aiyabot#268] [Ideas]
Hiya,

i'm setting up a server for me and my friends to mess around with.

TL;DR: I have Loras / embeddings that i'd like to be restricted, so for example only admins can view it because they might be still being tested or simply shouldn't be made available because people can't behave.

So my idea was to be able to tag Loras as restricted, making them hidden from the selection.

Additionally, preventing users to edit settings would be cool, i haven't found anything about this documented... so maybe i'm just silly.
Thank you!

**Comment by wizz13150** [comment]
Hey @LumiWasTaken 👋

You can add a simple condition to check if the User using /Settings is 'Moderator', or any role name to your convenience.

Just edit the line :
https://github.com/Kilvoctu/aiyabot/blob/4b3b62f3065f6d458bea328abd8fa1e0633d7a48/core/settingscog.py#L244

To :
```
        # Check if the user has the 'Moderator' role
        if not any(role.name == 'Moderator' for role in ctx.author.roles):
            await ctx.send_response('Sorry, you need the "Moderator" Role to use the settings command (:', ephemeral=True)
            return
            
        # get the channel id and check if a settings file exists
```
This will check if the discord user is member of the named role, and basic users won't be able to use the command in any channel.<br>

__________________________________________________
<br>
Regarding the Loras, it may be a more evolded edit regarding what you exactly want. Different approaches comes to my mind. 

The simple one is that you'll have to manually tag the Lora, from the bot or from the webui.
If you create 2 folders in the Lora folder for the webui, like : 

- Lora/Public
- Lora/Private

Or directly in the filename of the Loras.

You can simply parse the Loras names from the bot to check with a few lines, like above, if the user is allowed to use them.
Then the simple solution is, like above, to just reject the request.
Another solution is to just ignore the Lora.
Then others solutions comes to my mind. Depends..

A very simple solution, for me, is to edit : 
https://github.com/Kilvoctu/aiyabot/blob/4b3b62f3065f6d458bea328abd8fa1e0633d7a48/core/stablecog.py#L358-L360

To something like :
```
        if extra_net is not None:
            # Split the extra networks by spaces or commas, adjust as necessary
            extra_nets = extra_net.split()
            filtered_extra_nets = []

            for net in extra_nets:
                if net.lower().startswith('private'):
                    # Check if the user has the 'Moderator' role
                    if not any(role.name == 'Moderator' for role in ctx.author.roles):
                        await ctx.send_response('Sorry, you need the "Moderator" Role to use private networks.', ephemeral=True)
                        continue  # Skip adding this network
                # Add the network to the filtered list if it's allowed
                filtered_extra_nets.append(net)

            # Reconstruct the extra_net string from the filtered list
            extra_net = ' '.join(filtered_extra_nets)

            prompt, extra_net, net_multi = settings.extra_net_check(prompt, extra_net, net_multi)

        prompt = settings.extra_net_defaults(prompt, channel)
```

Here, if one of the selected Loras start with '**private**', it will check if the user has the '**Moderator**' discord role.
If the user doesn't have the role, only add the 'public' Loras to the prompt and ignore the 'private' ones. So the command continue.
Of course it's to adjust, probably not the best approach, tho. But it should do the job.
Also, the private Loras will also still be displayed for all users in the bot. 

To actually hide Loras for these users, as you want, would also require to change the 'autocomplete' when loading the extranet list to something else using a whitelist/blacklist. So a refactored function in settings.py and the lists in the config file. This way, only the lists need to be maintained.

- **NB: My bot is in use, i can't test this code rn, i'm not sure how to split `extra_net`**, but the big lines are here.

Cheers ! 🥂
